test/common/TestCommon.hs

```diff
@@ -0,0 +1,179 @@
+{-# LANGUAGE DataKinds, RankNTypes, TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+module TestCommon ((+\), (*\), (**\),
+                   listsToParameters,
+                   cmpTwo, cmpTwoSimple,
+                   qcPropDom, quickCheckTest0, fquad, quad
+                  ) where
+
+import Prelude
+
+import qualified Data.Array.DynamicS as OT
+import qualified Data.Array.ShapedS as OS
+import qualified Data.Vector.Generic as V
+import           Numeric.LinearAlgebra (Vector)
+import qualified Numeric.LinearAlgebra as HM
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+import HordeAd hiding (sumElementsVectorOfDual)
+
+(+\) :: DualMonad d r m
+     => DualNumber d r -> DualNumber d r -> m (DualNumber d r)
+(+\) u v = returnLet $ u + v
+
+(*\) :: DualMonad d r m
+     => DualNumber d r -> DualNumber d r -> m (DualNumber d r)
+(*\) u v = returnLet $ u * v
+
+(**\) :: DualMonad d r m
+      => DualNumber d r -> DualNumber d r -> m (DualNumber d r)
+(**\) u v = returnLet $ u ** v
+
+-- Checks if 2 numbers are close enough.
+close1 :: Double -> Double -> Bool
+close1 a b = abs (a - b) <= 1e-4
+
+-- Checks if 2 number pairs are close enough.
+close2 :: (Double, Double) -> (Double, Double) -> Property
+close2 (a1, b1) (a2, b2) = (close1 a1 a2) .&&. (close1 b1 b2)
+
+quad :: DualMonad d r m
+     => DualNumber d r -> DualNumber d r -> m (DualNumber d r)
+quad x y = do
+  x2 <- returnLet $ square x
+  y2 <- y *\ y
+  tmp <- x2 +\ y2
+  tmp +\ 5
+
+fquad :: forall r d m. DualMonad d r m
+      => DualNumberVariables d r -> m (DualNumber d r)
+fquad variables = do
+  let x = var0 variables 0
+      y = var0 variables 1
+  quad x y
+
+listsToParameters :: ([Double], [Double]) -> Domains Double
+listsToParameters (a0, a1) =
+  (V.fromList a0, V.singleton $ V.fromList a1, V.empty, V.empty)
+
+listsToParameters4 :: ([Double], [Double], [Double], [Double]) -> Domains Double
+listsToParameters4 (a0, a1, a2, aX) =
+  ( V.fromList a0
+  , V.singleton $ V.fromList a1
+  , if null a2 then V.empty else V.singleton $ HM.matrix 1 a2
+  , if null aX then V.empty else V.singleton $ OT.fromList [length aX] aX )
+
+quickCheckTest0 :: TestName
+       -> (forall d r m. ( DualMonad d r m
+                         , Floating (Out (DualNumber d (Vector r)))
+                         , Floating (Out (DualNumber d (OS.Array '[2] r))) )
+           => DualNumberVariables d r -> m (DualNumber d r))
+       -> ((Double, Double, Double) -> ([Double], [Double], [Double], [Double]))
+       -> TestTree
+quickCheckTest0 txt f fArg =
+  qcTestRanges txt f (listsToParameters4 . fArg) ((-2, -2, -2), (2, 2, 2)) ((-1e-7, -1e-7, -1e-7), (1e-7, 1e-7, 1e-7)) (-10, 10)
+
+-- A quick check to compare the derivatives and values of 2 given functions.
+cmpTwo :: (forall d r m. ( DualMonad d r m
```

Comment:
If we generalize from `Double` to `Num r` or `Floating r`, we can probably merge the two foralls and move them to the start of this type signature.

Reply:
A simple fix like that doesn't seem to work, due to existing `r ~ Double` or similar bindings, e.g. here: https://github.com/Mikolaj/horde-ad/blob/91dc3840ee27cc75ce1b7c5a1411600eb66bfe94/test/common/TestMnistCNN.hs#L623 .

Reply:
You are right, this is not trivial and out of scope of this PR. I had to do the following:

```hs
import           HordeAd.Core.DualClass (Dual)

cmpTwo
  :: ( m ~ DualMonadForward r, d ~ 'DModeDerivative, Dual d r ~ r
     , DualMonad d r m )
  => (DualNumberVariables d r -> m (DualNumber d r))
  -> (DualNumberVariables d r -> m (DualNumber d r))
  -> Domains r
  -> Domains r
  -> Domains r
  -> Domains r
  -> Property

cmpTwoSimple
  :: ( m ~ DualMonadForward r, d ~ 'DModeDerivative, Dual d r ~ r
     , DualMonad d r m )
  => (DualNumberVariables d r -> m (DualNumber d r))
  -> (DualNumberVariables d r -> m (DualNumber d r))
  -> Domains r
  -> Domains r
  -> Property
```

Reply:
Gladly imported, but in the other cases this didn't work. Shall we have a ticket to similarly simplify some of those type signatures?